### PR TITLE
drivers: platform: linux: Add axi_io implementation

### PR DESCRIPTION
**Pull request by dbogdan**
Implementation of AXI IO through UIO.

Signed-off-by: Dragos Bogdan <dragos.bogdan@analog.com>

### CODE REVIEW

**Comment by dbogdan** [comment]
v2:
   - if (uio_fd < 1) -> if (uio_fd < 0)

File: drivers/platform/linux/axi_io.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t axi_io_read(uint32_t base, uint32_t offset, uint32_t *data)
+{
+	char buf[16];
+	int ret;
+	int uio_fd;
+	void *uio_addr;
+
+	sprintf(buf, "/dev/uio%d", base);
```
Comment (mchindri) [inline drivers/platform/linux/axi_io.c]: the max uint32_t is 10 character, if this number is given will be a buffer overflow. Is the input is from some of our code then maybe it is ok if it is not given a big number. Same for write function.
Also, I think is good practice to use %u for unsigned. On small number is ok, but if the most significant bit is set %d will give a negative number.

File: drivers/platform/linux/axi_io.c [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+	uio_fd = open(buf, O_RDWR);
+	if (uio_fd < 0) {
+		printf("%s: Can't open %s\n\r", __func__, buf);
+		return FAILURE;
+	}
+
+	uio_addr = mmap(NULL,
```
Comment (mchindri) [inline drivers/platform/linux/axi_io.c]: shouldn't the returned value be checked?

File: drivers/platform/linux/axi_io.c [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			0x1000,
```
Comment (mchindri) [inline drivers/platform/linux/axi_io.c]: I think if you map from the offset address you can map only the needed bytes

File: drivers/platform/linux/axi_io.c [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+			PROT_READ|PROT_WRITE,
+			MAP_SHARED,
+			uio_fd,
+			0);
+
+	*data = (*((unsigned *) (uio_addr + offset)));
```
Comment (adrimbarean) [inline drivers/platform/linux/axi_io.c]: Changing this to
`*data = (*((uint32_t *)uio_addr + offset));`
should solve the Codancy issue.

File: drivers/platform/linux/axi_io.c [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	}
+
+	uio_addr = mmap(NULL,
+			0x1000,
+			PROT_READ|PROT_WRITE,
+			MAP_SHARED,
+			uio_fd,
+			0);
+
+	*((unsigned *) (uio_addr + offset)) = data;
```
Comment (adrimbarean) [inline drivers/platform/linux/axi_io.c]: Changing this to
`*((uint32_t *)uio_addr + offset) = data;`
should solve the Codancy issue.

File: drivers/platform/linux/axi_io.c [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t axi_io_write(uint32_t base, uint32_t offset, uint32_t data)
+{
+	char buf[16];
+	int ret;
+	int uio_fd;
+	void *uio_addr;
+
+	sprintf(buf, "/dev/uio%d", base);
```
Comment (adrimbarean) [inline drivers/platform/linux/axi_io.c]: Changing this to
`sprintf(buf, "/dev/uio%u", base);`
should solve the Codancy issue.

File: drivers/platform/linux/axi_io.c (hunk 2 above)
Comment (dbogdan) [reply drivers/platform/linux/axi_io.c]: Fixed that.

**Comment by dbogdan** [comment]
v2:
   - use %"PRIu32" instead of %d for printing uint32_t base;
   - check mmap() return value;
   - cast data to uint32_t instead of unsigned.

**Comment by dbogdan** [comment]
v4:
   - create axi_io_read_write() in order to remove the code duplication;
   - use munmap() for removing the mappings created by mmap();
   - increase buf[] length to 32.

**Comment by dbogdan** [comment]
v5:
   - Fixed Codacy complaints.

**Comment by dbogdan** [comment]
v6:
   - Fix mmap length value (it should take in consideration the offset too).

**Comment by dbogdan** [comment]
v7:
   - fix Codacy complaints.

File: drivers/platform/linux/axi_io.c [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <fcntl.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include "error.h"
+#include "axi_io.h"
+
```
Comment (mchindri) [inline drivers/platform/linux/axi_io.c]: You forgot the function definition header

Comment (dbogdan) [reply drivers/platform/linux/axi_io.c]: Done.

**Pull request merged**